File: VB/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System.Web.Mvc
Imports DevExpress.Razor.Models
Imports DevExpress.Web.Mvc

Namespace DevExpress.Razor.Controllers
	Public Class HomeController
		Inherits Controller
		Public Function Index() As ActionResult
			Return View(Company.OfficeRooms)
		End Function
		Public Function GridViewMasterPartial() As ActionResult
			Return PartialView("GridViewMasterPartial", Company.OfficeRooms)
		End Function
		Public Function GridViewMasterAddNewPartial(<ModelBinder(GetType(DevExpressEditorsBinder))> ByVal room As Room) As ActionResult
			If ModelState.IsValid Then
				Company.InsertRoom(room)
			End If
			Return PartialView("GridViewMasterPartial", Company.OfficeRooms)
		End Function
		Public Function GridViewMasterUpdatePartial(<ModelBinder(GetType(DevExpressEditorsBinder))> ByVal room As Room) As ActionResult
			If ModelState.IsValid Then
				Company.UpdateRoom(room)
			End If
			Return PartialView("GridViewMasterPartial", Company.OfficeRooms)
		End Function
		Public Function GridViewMasterDeletePartial(ByVal ID As Integer) As ActionResult
			Company.RemoveRoomByID(ID)
			Return PartialView("GridViewMasterPartial", Company.OfficeRooms)
		End Function

		Public Function GridViewDetailPartial(ByVal roomID As Integer) As ActionResult
			ViewData("RoomID") = roomID
			Return PartialView("GridViewDetailPartial", Company.GetPersonsByRoom(roomID))
		End Function
		Public Function GridViewDetailAddNewPartial(<ModelBinder(GetType(DevExpressEditorsBinder))> ByVal person As Person, ByVal roomID As Integer) As ActionResult
			ViewData("RoomID") = roomID
			person.RoomID = roomID
			If ModelState.IsValid Then
				Company.InsertPerson(person)
			End If
			Return PartialView("GridViewDetailPartial", Company.GetPersonsByRoom(roomID))
		End Function
		Public Function GridViewDetailUpdatePartial(<ModelBinder(GetType(DevExpressEditorsBinder))> ByVal person As Person, ByVal roomID As Integer) As ActionResult
			ViewData("RoomID") = roomID
			If ModelState.IsValid Then
				Company.UpdatePerson(person)
			End If
			Return PartialView("GridViewDetailPartial", Company.GetPersonsByRoom(roomID))
		End Function
		Public Function GridViewDetailDeletePartial(ByVal ID As Integer, ByVal roomID As Integer) As ActionResult
			ViewData("RoomID") = roomID
			Company.RemovePersonByID(ID)
			Return PartialView("GridViewDetailPartial", Company.GetPersonsByRoom(roomID))
		End Function
	End Class
End Namespace

File: CS/DevExpress.Razor/Controllers/HomeController.cs
using System.Web.Mvc;
using DevExpress.Razor.Models;
using DevExpress.Web.Mvc;

namespace DevExpress.Razor.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View(Company.OfficeRooms);
        }
        public ActionResult GridViewMasterPartial() {
            return PartialView("GridViewMasterPartial", Company.OfficeRooms);
        }
        public ActionResult GridViewMasterAddNewPartial([ModelBinder(typeof(DevExpressEditorsBinder))] Room room) {
            if (ModelState.IsValid)
                Company.InsertRoom(room);
            return PartialView("GridViewMasterPartial", Company.OfficeRooms);
        }
        public ActionResult GridViewMasterUpdatePartial([ModelBinder(typeof(DevExpressEditorsBinder))] Room room) {
            if (ModelState.IsValid)
                Company.UpdateRoom(room);
            return PartialView("GridViewMasterPartial", Company.OfficeRooms);
        }
        public ActionResult GridViewMasterDeletePartial(int ID){
            Company.RemoveRoomByID(ID);
            return PartialView("GridViewMasterPartial", Company.OfficeRooms);
        }

        public ActionResult GridViewDetailPartial(int roomID) {
            ViewData["RoomID"] = roomID;
            return PartialView("GridViewDetailPartial", Company.GetPersonsByRoom(roomID));
        }
        public ActionResult GridViewDetailAddNewPartial([ModelBinder(typeof(DevExpressEditorsBinder))] Person person, int roomID) {
            ViewData["RoomID"] = roomID;
            person.RoomID = roomID;
            if(ModelState.IsValid)
                Company.InsertPerson(person);
            return PartialView("GridViewDetailPartial", Company.GetPersonsByRoom(roomID));
        }
        public ActionResult GridViewDetailUpdatePartial([ModelBinder(typeof(DevExpressEditorsBinder))] Person person, int roomID) {
            ViewData["RoomID"] = roomID;
            if(ModelState.IsValid)
                Company.UpdatePerson(person);
            return PartialView("GridViewDetailPartial", Company.GetPersonsByRoom(roomID));
        }
        public ActionResult GridViewDetailDeletePartial(int ID, int roomID){
            ViewData["RoomID"] = roomID;
            Company.RemovePersonByID(ID);
            return PartialView("GridViewDetailPartial", Company.GetPersonsByRoom(roomID));
        }
    }
}